File: crates/defi-cli/src/commands/compare.rs
use clap::Args;
use std::process::Command;

use defi_core::error::Result;
use defi_core::registry::{ProtocolCategory, Registry};

use crate::output::OutputMode;

#[derive(Args)]
pub struct CompareArgs {
    /// Asset to compare yields for (e.g., USDC, ETH, BTC)
    #[arg(long, default_value = "USDC")]
    pub asset: String,

    /// Include perp funding rates from perp-cli
    #[arg(long, default_value_t = true)]
    pub perps: bool,

    /// Include lending rates from defi-cli
    #[arg(long, default_value_t = true)]
    pub lending: bool,

    /// Minimum absolute APY to show
    #[arg(long, default_value = "1.0")]
    pub min_apy: f64,
}

pub async fn run(args: CompareArgs, registry: &Registry, output: &OutputMode) -> Result<()> {
    let start = std::time::Instant::now();
    let mut opportunities: Vec<serde_json::Value> = Vec::new();

    // 1. Collect perp funding rates (call perp-cli)
    if args.perps
        && let Ok(perp_data) = fetch_perp_rates().await
    {
        for opp in perp_data {
            let apy = opp["apy"].as_f64().unwrap_or(0.0);
            if apy.abs() >= args.min_apy {
                opportunities.push(opp);
            }
        }
    }

    // 2. Collect lending rates across all chains (parallel)
    if args.lending {
        let lending_data = fetch_lending_rates(registry, &args.asset).await;
        for opp in lending_data {
            let apy = opp["apy"].as_f64().unwrap_or(0.0);
            if apy.abs() >= args.min_apy {
                opportunities.push(opp);
            }
        }
    }

    // Sort by absolute APY descending
    opportunities.sort_by(|a, b| {
        b["apy"]
            .as_f64()
            .unwrap_or(0.0)
            .abs()
            .partial_cmp(&a["apy"].as_f64().unwrap_or(0.0).abs())
            .unwrap()
    });

    let scan_ms = start.elapsed().as_millis();

    let result = serde_json::json!({
        "asset": args.asset,
        "scan_duration_ms": scan_ms,
        "total_opportunities": opportunities.len(),
        "opportunities": opportunities,
    });

    output.print(&result)?;
    Ok(())
}

/// Fetch perp funding rates by calling `perp --json arb scan --rates`
async fn fetch_perp_rates() -> std::result::Result<Vec<serde_json::Value>, String> {
    let output = Command::new("perp")
        .args(["--json", "arb", "scan", "--rates"])
        .output()
        .map_err(|e| format!("perp-cli not found: {e}"))?;

    if !output.status.success() {
        return Err("perp arb scan failed".into());
    }

    let data: serde_json::Value =
        serde_json::from_slice(&output.stdout).map_err(|e| format!("perp JSON parse: {e}"))?;

    let symbols = data
        .get("data")
        .and_then(|d| d.get("symbols"))
        .or_else(|| data.get("symbols"))
        .and_then(|s| s.as_array());

    let mut results = Vec::new();

    if let Some(symbols) = symbols {
        for sym in symbols {
            let symbol = sym["symbol"].as_str().unwrap_or("?");
            let max_spread = sym["maxSpreadAnnual"].as_f64().unwrap_or(0.0);
            let long_ex = sym["longExchange"].as_str().unwrap_or("?");
            let short_ex = sym["shortExchange"].as_str().unwrap_or("?");

            if max_spread.abs() > 0.0 {
                results.push(serde_json::json!({
                    "type": "perp_funding",
                    "asset": symbol,
                    "apy": round2(max_spread),
                    "detail": format!("long {} / short {}", long_ex, short_ex),
                    "risk": if max_spread.abs() > 50.0 { "high" } else if max_spread.abs() > 20.0 { "medium" } else { "low" },
                    "source": "perp-cli",
                }));
            }

            // Also add individual exchange rates
            if let Some(rates) = sym["rates"].as_array() {
                for rate in rates {
                    let exchange = rate["exchange"].as_str().unwrap_or("?");
                    let annual = rate["annualizedPct"].as_f64().unwrap_or(0.0);
                    if annual.abs() > 1.0 {
                        results.push(serde_json::json!({
                            "type": "perp_rate",
                            "asset": symbol,
                            "apy": round2(annual),
                            "detail": exchange,
                            "risk": if annual.abs() > 50.0 { "high" } else if annual.abs() > 20.0 { "medium" } else { "low" },
                            "source": "perp-cli",
                        }));
                    }
                }
            }
        }
    }

    Ok(results)
}

/// Fetch lending rates across all chains
async fn fetch_lending_rates(registry: &Registry, asset: &str) -> Vec<serde_json::Value> {
    let chain_keys: Vec<String> = registry.chains.keys().cloned().collect();
    let mut join_set = tokio::task::JoinSet::new();

    for ck in &chain_keys {
        let chain = match registry.get_chain(ck) {
            Ok(c) => c,
            Err(_) => continue,
        };
        let asset_addr = match registry.resolve_token(ck, asset) {
            Ok(t) => t.address,
            Err(_) => continue,
        };
        let protos: Vec<_> = registry
            .get_protocols_for_chain(ck)
            .iter()
            .filter(|p| p.category == ProtocolCategory::Lending && p.interface == "aave_v3")
            .cloned()
            .cloned()
            .collect();

        if protos.is_empty() {
            continue;
        }

        let chain_name = chain.name.clone();
        let rpc = chain.effective_rpc_url();
        let asset_sym = asset.to_string();

        join_set.spawn(async move {
            let mut rates = Vec::new();
            for proto in &protos {
                if let Ok(lending) =
                    defi_protocols::factory::create_lending_with_rpc(proto, Some(&rpc))
                    && let Ok(r) = lending.get_rates(asset_addr).await
                    && r.supply_apy > 0.0
                {
                    rates.push(serde_json::json!({
                        "type": "lending_supply",
                        "asset": asset_sym,
                        "apy": round2(r.supply_apy),
                        "detail": format!("{} ({})", r.protocol, chain_name),
                        "risk": "low",
                        "source": "defi-cli",
                    }));
                }
            }
            rates
        });
    }

    let mut all = Vec::new();
    while let Some(result) = join_set.join_next().await {
        if let Ok(rates) = result {
            all.extend(rates);
        }
    }
    all
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

File: ts/packages/defi-cli/src/commands/compare.ts
import { spawnSync } from "child_process";
import type { Command } from "commander";
import { Registry, ProtocolCategory } from "@hypurrquant/defi-core";
import { createLending } from "@hypurrquant/defi-protocols";
import type { OutputMode } from "../output.js";
import { printOutput } from "../output.js";

function round2(x: number): number {
  return Math.round(x * 100) / 100;
}

/** Fetch perp funding rates by calling `perp --json arb scan --rates` */
async function fetchPerpRates(): Promise<unknown[]> {
  // Try global `perp` first, then npx fallback
  let result = spawnSync("perp", ["--json", "arb", "scan", "--rates"], { encoding: "utf8", timeout: 30000 });
  if (result.error || result.status !== 0) {
    result = spawnSync("npx", ["-y", "perp-cli@latest", "--json", "arb", "scan", "--rates"], {
      encoding: "utf8",
      timeout: 60000,
    });
  }

  if (result.error || result.status !== 0) {
    throw new Error("perp-cli not found or failed");
  }

  let data: unknown;
  try {
    data = JSON.parse(result.stdout);
  } catch {
    throw new Error("perp JSON parse error");
  }

  const d = data as Record<string, unknown>;
  const symbolsRaw =
    (d["data"] as Record<string, unknown> | undefined)?.["symbols"] ??
    d["symbols"];
  const symbols = Array.isArray(symbolsRaw) ? symbolsRaw : [];

  const results: unknown[] = [];
  for (const sym of symbols as Array<Record<string, unknown>>) {
    const symbol = (sym["symbol"] as string) ?? "?";
    const maxSpread = (sym["maxSpreadAnnual"] as number) ?? 0;
    const longEx = (sym["longExchange"] as string) ?? "?";
    const shortEx = (sym["shortExchange"] as string) ?? "?";

    if (Math.abs(maxSpread) > 0) {
      results.push({
        type: "perp_funding",
        asset: symbol,
        apy: round2(maxSpread),
        detail: `long ${longEx} / short ${shortEx}`,
        risk: Math.abs(maxSpread) > 50 ? "high" : Math.abs(maxSpread) > 20 ? "medium" : "low",
        source: "perp-cli",
      });
    }

    const rates = Array.isArray(sym["rates"]) ? (sym["rates"] as Array<Record<string, unknown>>) : [];
    for (const rate of rates) {
      const exchange = (rate["exchange"] as string) ?? "?";
      const annual = (rate["annualizedPct"] as number) ?? 0;
      if (Math.abs(annual) > 1.0) {
        results.push({
          type: "perp_rate",
          asset: symbol,
          apy: round2(annual),
          detail: exchange,
          risk: Math.abs(annual) > 50 ? "high" : Math.abs(annual) > 20 ? "medium" : "low",
          source: "perp-cli",
        });
      }
    }
  }

  return results;
}

/** Fetch lending supply rates across all chains for a given asset */
async function fetchLendingRates(registry: Registry, asset: string): Promise<unknown[]> {
  const chainKeys = Array.from(registry.chains.keys());

  const tasks = chainKeys.map(async (ck) => {
    try {
      const chain = registry.getChain(ck);
      const chainName = chain.name.toLowerCase();

      let assetAddr: `0x${string}`;
      try {
        assetAddr = registry.resolveToken(chainName, asset).address;
      } catch {
        return [];
      }

      const protos = registry
        .getProtocolsForChain(chainName)
        .filter((p) => p.category === ProtocolCategory.Lending && p.interface === "aave_v3");

      if (protos.length === 0) return [];

      const rpc = chain.effectiveRpcUrl();
      const rates: unknown[] = [];

      for (const proto of protos) {
        try {
          const lending = createLending(proto, rpc);
          const r = await lending.getRates(assetAddr);
          if (r.supply_apy > 0) {
            rates.push({
              type: "lending_supply",
              asset,
              apy: round2(r.supply_apy * 100),
              detail: `${r.protocol} (${chain.name})`,
              risk: "low",
              source: "defi-cli",
            });
          }
        } catch {
          // skip unavailable protocols
        }
      }

      return rates;
    } catch {
      return [];
    }
  });

  const nested = await Promise.all(tasks);
  return nested.flat();
}

export function registerCompare(parent: Command, getOpts: () => OutputMode): void {
  parent
    .command("compare")
    .description("Compare all yield sources: perp funding vs lending APY vs staking")
    .option("--asset <token>", "Token symbol to compare (e.g. USDC, ETH)", "USDC")
    .option("--no-perps", "Exclude perp funding rates")
    .option("--no-lending", "Exclude lending rates")
    .option("--min-apy <pct>", "Minimum absolute APY to show", "1.0")
    .action(async (opts) => {
      try {
        const registry = Registry.loadEmbedded();
        const asset: string = opts.asset ?? "USDC";
        const includePerps: boolean = opts.perps !== false;
        const includeLending: boolean = opts.lending !== false;
        const minApy = parseFloat(opts.minApy ?? "1.0");

        const t0 = Date.now();
        const opportunities: unknown[] = [];

        // 1. Perp funding rates
        if (includePerps) {
          try {
            const perpData = await fetchPerpRates();
            for (const opp of perpData) {
              const apy = Math.abs((opp as Record<string, unknown>)["apy"] as number ?? 0);
              if (apy >= minApy) opportunities.push(opp);
            }
          } catch {
            // perp-cli not available — skip silently
          }
        }

        // 2. Lending rates across all chains
        if (includeLending) {
          const lendingData = await fetchLendingRates(registry, asset);
          for (const opp of lendingData) {
            const apy = Math.abs((opp as Record<string, unknown>)["apy"] as number ?? 0);
            if (apy >= minApy) opportunities.push(opp);
          }
        }

        // Sort by absolute APY descending
        opportunities.sort((a, b) => {
          const aApy = Math.abs((a as Record<string, unknown>)["apy"] as number ?? 0);
          const bApy = Math.abs((b as Record<string, unknown>)["apy"] as number ?? 0);
          return bApy - aApy;
        });

        const scanMs = Date.now() - t0;

        printOutput(
          {
            asset,
            scan_duration_ms: scanMs,
            total_opportunities: opportunities.length,
            opportunities,
          },
          getOpts(),
        );
      } catch (err) {
        printOutput({ error: String(err) }, getOpts());
        process.exit(1);
      }
    });
}
